File: Behavior/Iterator/Cpp/iterator.cpp
#include <iostream>
#include <vector>

// 迭代器基类
template <typename T>
class Iterator
{
public:
    virtual ~Iterator() {}
    virtual T next() = 0;
    virtual bool hasNext() const = 0;
};

// 容器接口
template <typename T>
class Container
{
public:
    virtual ~Container() {}
    virtual Iterator<T> *createIterator() const = 0;
    // 其他容器相关方法
};

// 具体迭代器
template <typename T>
class ConcreteIterator : public Iterator<T>
{
private:
    const std::vector<T> &collection;
    size_t current = 0;

public:
    ConcreteIterator(const std::vector<T> &collection) : collection(collection) {}

    T next() override
    {
        return collection[current++];
    }

    bool hasNext() const override
    {
        return current < collection.size();
    }
};

// 具体容器
template <typename T>
class ConcreteContainer : public Container<T>
{
private:
    std::vector<T> elements;

public:
    void add(T element)
    {
        elements.push_back(element);
    }

    Iterator<T> *createIterator() const override
    {
        return new ConcreteIterator<T>(elements);
    }

    // 可以添加其他容器相关的方法
};

int main()
{
    ConcreteContainer<int> container;
    container.add(1);
    container.add(2);
    container.add(3);

    Iterator<int> *it = container.createIterator();
    while (it->hasNext())
    {
        std::cout << it->next() << std::endl;
    }

    delete it; // 注意：使用原始指针需要手动管理内存
    return 0;
}

File: Behavior/Iterator/Go/iterator.go
package main

import "fmt"

// Iterator 是迭代器接口
type Iterator interface {
	HasNext() bool
	Next() interface{}
}

// IntSliceIterator 是int切片的迭代器
type IntSliceIterator struct {
	slice []int
	index int
}

// HasNext 方法检查迭代器是否已到达切片的末尾
func (i *IntSliceIterator) HasNext() bool {
	return i.index < len(i.slice)
}

// Next 方法返回切片中的下一个元素，并将索引向前移动
func (i *IntSliceIterator) Next() interface{} {
	if i.HasNext() {
		value := i.slice[i.index]
		i.index++
		return value
	}
	return nil // 或者返回错误
}

// IterableIntSlice 是一个可以迭代的int切片
type IterableIntSlice []int

// CreateIterator 创建一个IntSliceIterator来遍历切片
func (s IterableIntSlice) CreateIterator() Iterator {
	return &IntSliceIterator{
		slice: s,
		index: 0,
	}
}

func main() {
	slice := IterableIntSlice{1, 2, 3, 4, 5}
	it := slice.CreateIterator()
	for it.HasNext() {
		value := it.Next()
		fmt.Println(value)
	}
}
